File: implementations/kotlin/src/main/kotlin/Zobrist.kt
// Zobrist hashing for position detection

object Zobrist {
    private val pieceKeys = Array(64 * 12) { 0L }
    private var turnKey = 0L
    private val castlingKeys = Array(16) { 0L }
    private val enPassantKeys = Array(64) { 0L }
    private var initialized = false

    private class Xorshift64(private var state: Long) {
        fun next(): Long {
            state = state xor (state shl 13)
            state = state xor (state ushr 7)
            state = state xor (state shl 17)
            return state
        }
    }

    fun init() {
        if (initialized) return
        val rng = Xorshift64(0x123456789ABCDEF0L)

        for (i in pieceKeys.indices) pieceKeys[i] = rng.next()
        turnKey = rng.next()
        for (i in castlingKeys.indices) castlingKeys[i] = rng.next()
        for (i in enPassantKeys.indices) enPassantKeys[i] = rng.next()
        
        initialized = true
    }

    private fun getPieceIndex(type: PieceType, color: Color): Int {
        val base = when (type) {
            PieceType.PAWN -> 0
            PieceType.KNIGHT -> 1
            PieceType.BISHOP -> 2
            PieceType.ROOK -> 3
            PieceType.QUEEN -> 4
            PieceType.KING -> 5
        }
        return if (color == Color.WHITE) base else base + 6
    }

    fun computeHash(gameState: GameState): Long {
        init()
        var h = 0L
        
        for (i in 0..63) {
            val piece = gameState.board[i]
            if (piece != null) {
                h = h xor pieceKeys[i * 12 + getPieceIndex(piece.type, piece.color)]
            }
        }
        
        if (gameState.turn == Color.BLACK) {
            h = h xor turnKey
        }
        
        var castlingIndex = 0
        if (gameState.castlingRights.whiteKingside) castlingIndex = castlingIndex or 1
        if (gameState.castlingRights.whiteQueenside) castlingIndex = castlingIndex or 2
        if (gameState.castlingRights.blackKingside) castlingIndex = castlingIndex or 4
        if (gameState.castlingRights.blackQueenside) castlingIndex = castlingIndex or 8
        h = h xor castlingKeys[castlingIndex]
        
        val ep = gameState.enPassantTarget
        if (ep != null) {
            h = h xor enPassantKeys[ep]
        }
        
        return h
    }
}

File: implementations/go/zobrist.go
package main

var (
	pieceKeys      [64][12]uint64
	sideToMoveKey  uint64
	castlingKeys   [16]uint64
	enPassantKeys  [8]uint64
	zobristInitial bool
)

type Xorshift64 struct {
	state uint64
}

func (x *Xorshift64) Next() uint64 {
	x.state ^= x.state << 13
	x.state ^= x.state >> 7
	x.state ^= x.state << 17
	return x.state
}

func initZobrist() {
	if zobristInitial {
		return
	}

	rng := &Xorshift64{state: 0x123456789ABCDEF0}

	for i := 0; i < 64; i++ {
		for j := 0; j < 12; j++ {
			pieceKeys[i][j] = rng.Next()
		}
	}

	sideToMoveKey = rng.Next()

	for i := 0; i < 16; i++ {
		castlingKeys[i] = rng.Next()
	}

	for i := 0; i < 8; i++ {
		enPassantKeys[i] = rng.Next()
	}

	zobristInitial = true
}

func getPieceIndex(p Piece) int {
	base := int(p.Type) - 1
	if p.Color == Black {
		base += 6
	}
	return base
}

func computeZobristHash(gs *GameState) uint64 {
	initZobrist()
	var h uint64

	for r := 0; r < 8; r++ {
		for f := 0; f < 8; f++ {
			p := gs.Board[r][f]
			if p.Type != Empty {
				h ^= pieceKeys[r*8+f][getPieceIndex(p)]
			}
		}
	}

	if gs.ActiveColor == Black {
		h ^= sideToMoveKey
	}

	castlingIndex := 0
	if gs.CastlingRights[White][KingsideCastle] {
		castlingIndex |= 1
	}
	if gs.CastlingRights[White][QueensideCastle] {
		castlingIndex |= 2
	}
	if gs.CastlingRights[Black][KingsideCastle] {
		castlingIndex |= 4
	}
	if gs.CastlingRights[Black][QueensideCastle] {
		castlingIndex |= 8
	}
	h ^= castlingKeys[castlingIndex]

	if gs.EnPassantTarget != nil {
		h ^= enPassantKeys[gs.EnPassantTarget.File]
	}

	return h
}